pkg/db/db_releases.go

```diff
@@ -0,0 +1,681 @@
+	rows, err := tx.QueryContext(
+		ctx,
+		selectQuery,
+	)
+
+	return h.processReleaseManifestRows(ctx, err, rows)
+}
+
+func (h *DBHandler) DBSelectReleasesByAppLatestEslVersion(ctx context.Context, tx *sql.Tx, app string, ignorePrepublishes bool) ([]*DBReleaseWithMetaData, error) {
+	selectQuery := h.AdaptQuery(`
+	SELECT created, appName, metadata, manifests, releaseVersion, environments
+	FROM releases
+	WHERE appname=?
+	ORDER BY releaseversion DESC;`,
+	)
+	span, ctx := tracer.StartSpanFromContext(ctx, "DBSelectReleasesByAppLatestEslVersion")
+	defer span.Finish()
+	span.SetTag("query", selectQuery)
+	rows, err := tx.QueryContext(
+		ctx,
+		selectQuery,
+		app,
+	)
+
+	return h.processReleaseRows(ctx, err, rows, ignorePrepublishes, true)
+}
+
+func (h *DBHandler) DBSelectLatestReleaseOfApp(ctx context.Context, tx *sql.Tx, app string, ignorePrepublishes bool) (*DBReleaseWithMetaData, error) {
+	selectQuery := h.AdaptQuery(`
+	SELECT created, appName, metadata, releaseVersion, environments
+	FROM releases
+	WHERE appName=?
+	ORDER BY releaseVersion DESC
+	LIMIT 1;`)
+	span, ctx := tracer.StartSpanFromContext(ctx, "DBSelectLatestReleaseOfApp")
+	defer span.Finish()
+	span.SetTag("query", selectQuery)
+	span.SetTag("appName", app)
+	rows, err := tx.QueryContext(
+		ctx,
+		selectQuery,
+		app,
+	)
+
+	return h.processReleaseRow(ctx, err, rows, ignorePrepublishes, false)
+}
+
+func (h *DBHandler) DBSelectAllReleasesOfApp(ctx context.Context, tx *sql.Tx, app string) ([]int64, error) {
+	selectQuery := h.AdaptQuery(`SELECT releaseVersion
+	FROM releases
+	WHERE appName=?
+	ORDER BY releaseVersion;`)
+	span, ctx := tracer.StartSpanFromContext(ctx, "DBSelectAllReleasesOfApp")
+	defer span.Finish()
+	span.SetTag("query", selectQuery)
+	rows, err := tx.QueryContext(
+		ctx,
+		selectQuery,
+		app,
+	)
+	return h.processAppReleaseVersionsRows(ctx, err, rows)
+}
+
+func (h *DBHandler) DBSelectAllReleasesOfAllApps(ctx context.Context, tx *sql.Tx) (map[string][]int64, error) {
+	span, ctx := tracer.StartSpanFromContext(ctx, "DBSelectAllReleasesOfAllApps")
+	defer span.Finish()
+	selectQuery := h.AdaptQuery(`
+	SELECT appname, releaseVersion 
+	FROM releases;`)
+
+	span.SetTag("query", selectQuery)
+	rows, err := tx.QueryContext(
+		ctx,
+		selectQuery,
+	)
+	return h.processAllAppsReleaseVersionsRows(ctx, err, rows)
+}
+
+// INSERT, UPDATE, DELETES
+
+func (h *DBHandler) DBUpdateOrCreateRelease(ctx context.Context, transaction *sql.Tx, release DBReleaseWithMetaData) error {
+	span, ctx := tracer.StartSpanFromContext(ctx, "DBUpdateOrCreateRelease")
+	defer span.Finish()
+	retrievedRelease, err := h.DBSelectReleaseByVersion(ctx, transaction, release.App, release.ReleaseNumber, false)
+	if err != nil {
+		return err
+	}
+	if retrievedRelease == nil {
+		err = h.insertReleaseRow(ctx, transaction, release)
+		if err != nil {
+			return err
+		}
+	} else {
+		err = h.updateReleaseRow(ctx, transaction, release)
+		if err != nil {
+			return err
+		}
+	}
```

Comment:
This should probably be an "upsert" instead of 3 queries. Something like this:
```
insert into apps (created, appname, statechange, metadata)
VALUES (now(), 'echo-3', 'AppStateChangeModify', '')
ON CONFLICT(version, appname)
DO UPDATE Set statechange = EXCLUDED.statechange, metadata = EXCLUDED.metadata ;
```

Reply:
Done.